sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -441,15 +599,35 @@ public void validate(PBegin input) {
           "Kafka bootstrap servers should be set");
       checkArgument(getTopics().size() > 0 || getTopicPartitions().size() > 0,
           "Kafka topics or topic_partitions are required");
-      checkNotNull(getKeyCoder(), "Key coder must be set");
-      checkNotNull(getValueCoder(), "Value coder must be set");
+      checkNotNull(getKeyDeserializer(), "Key deserializer must be set");
+      checkNotNull(getValueDeserializer(), "Value deserializer must be set");
+
+      if (input != null) {
+        CoderRegistry registry = input.getPipeline().getCoderRegistry();
+        checkNotNull(getOrInferCoder(registry, getKeyDeserializer(), getKeyCoder()),
+                "Key coder must be set");
+        checkNotNull(getOrInferCoder(registry, getValueDeserializer(), getValueCoder()),
+                "Value coder must be set");
+      } else {
```

Comment:
Yes, I think so, but only some of them, such as `testUnboundedSourceSplits`, which manually calls`generateInitialSplits`  and `Read.from` on each split.  This means that `validate` is called through `Source.validate` which doesn't get a `PInput`